Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.h

```diff
@@ -236,5 +318,130 @@ namespace Terrain
+    inline AZStd::shared_ptr<AzFramework::Terrain::TerrainDataRequests::TerrainJobContext> TerrainSystem::ProcessFromListAsync(
+        SynchronousFunctionType synchronousFunction,
+        const AZStd::span<VectorType>& inPositions,
+        AzFramework::Terrain::SurfacePointListFillCallback perPositionCallback,
+        Sampler sampleFilter,
+        AZStd::shared_ptr<ProcessAsyncParams> params) const
+    {
+        // Determine the number of jobs to split the work into based on:
+        // 1. The number of available worker threads.
+        // 2. The desired number of jobs as passed in.
+        // 3. The number of positions being processed.
+        const int32_t numWorkerThreads = m_terrainJobManager->GetNumWorkerThreads();
+        const int32_t numJobsDesired = params ? params->m_desiredNumberOfJobs : ProcessAsyncParams::NumJobsDefault;
+        const int32_t numJobsMax = (numJobsDesired > 0) ? AZStd::min(numWorkerThreads, numJobsDesired) : numWorkerThreads;
+        const int32_t numPositionsToProcess = static_cast<int32_t>(inPositions.size());
+        const int32_t numJobs = AZStd::min(numJobsMax, numPositionsToProcess);
+        if (numJobs <= 0)
+        {
+            AZ_Warning("TerrainSystem", false, "No positions to process.");
+            return nullptr;
+        }
+
+        // Create a terrain job context to return so the jobs can be cancelled.
+        AZStd::shared_ptr<TerrainJobContext> jobContext = AZStd::make_shared<TerrainJobContext>(*m_terrainJobManager);
+
+        // Split the work across multiple jobs.
+        const int32_t numPositionsPerJob = numPositionsToProcess / numJobs;
+        AZStd::shared_ptr<AZStd::atomic_int> numJobCompletionsRemaining = AZStd::make_shared<AZStd::atomic_int>(numJobs);
+        for (int32_t i = 0; i < numJobs; ++i)
+        {
+            // If the number of positions can't be divided evenly by the number of jobs,
+            // ensure we still process the remaining positions along with the final job.
+            const size_t subSpanOffset = i * numPositionsPerJob;
+            const size_t subSpanCount = (i < numJobs - 1) ? numPositionsPerJob : AZStd::dynamic_extent;
+
+            // Define the job function using the sub span of positions to process.
+            const AZStd::span<VectorType>& positionsToProcess = inPositions.subspan(subSpanOffset, subSpanCount);
+            auto jobFunction = [=]()
+            {
+                // Process the sub span of positions, decrement the number of completions remaining,
+                // and invoke the completion callback if this happens to be the final job completed.
+                synchronousFunction(positionsToProcess, perPositionCallback, sampleFilter);
+                if (--(*numJobCompletionsRemaining) == 0 &&
+                    params && params->m_completionCallback &&
+                    !jobContext->IsCancelled())
+                {
+                    params->m_completionCallback();
+                }
```

Comment:
Yes, now that jobs can only be cancelled before they start, we will never get into this process job function is the request has been cancelled (because the JobManager will never start it in the first place). That actually made me realize that even with the previous behaviour where we were able to cancel jobs that had already started, if the request was cancelled before the job started then the callback would never have been invoked, which would be a problem because the calling code would be expecting the completion callback to get invoked in all circumstances.

So I changed it to explicitly state that the completion callback will only get called if the job actually completes; if the job does get cancelled while it's already in progress, the work will still happen, but the callback will not get called, so it is up to the caller to keep track of whether they cancel the job or not.

Reply:
I changed the way cancellation works such that we can now guarantee all terrain jobs will at least begin processing, so as part of that I changed this back to always call the completion callback regardless of whether the job was cancelled or not.